### PR TITLE
Modifation to allow mixed element types in matrix operations

### PR DESCRIPTION
**Pull request by awbsmith**
This should enable things like automatic differentiation:

``` Julia
using FixedSizeArrays
using ForwardDiff

R = rand(Mat{2,2,Float64})
X = Vec(ForwardDiff.GradientNumber(0.0, (1.0, 0.0)), ForwardDiff.GradientNumber(1.0, (0.0, 1.0)))

Y = R * X
```

As per [issue 93](https://github.com/SimonDanisch/FixedSizeArrays.jl/issues/93)

### CODE REVIEW

File: src/ops.jl [hunk 1]
```diff
@@ -115,21 +115,39 @@ end
...
 call(::BilinearDotFunctor, a, b) = a*b
 @inline bilindot{T <: Union{FixedArray, Tuple}}(a::T, b::T) = sum(map(DotFunctor(), a, b))
-@inline bilindot{T1 <: Tuple, T2 <: FixedArray}(a::T1, b::T2) = sum(map(DotFunctor(), a, b))
 
-@inline bilindot{T}(a::NTuple{1,T}, b::NTuple{1,T}) = @inbounds return a[1]*b[1]
-@inline bilindot{T}(a::NTuple{2,T}, b::NTuple{2,T}) = @inbounds return (a[1]*b[1] + a[2]*b[2])
-@inline bilindot{T}(a::NTuple{3,T}, b::NTuple{3,T}) = @inbounds return (a[1]*b[1] + a[2]*b[2] + a[3]*b[3])
-@inline bilindot{T}(a::NTuple{4,T}, b::NTuple{4,T}) = @inbounds return (a[1]*b[1] + a[2]*b[2] + a[3]*b[3]+a[4]*b[4])
+
+immutable BilinearDotFunctor <: Func{2} end
```
Comment (mschauer) [inline src/ops.jl]: Double line

**Comment by mschauer** [comment]
@c42f Will this be superseded by your work?

File: src/ops.jl [hunk 2]
```diff
@@ -115,21 +115,39 @@ end
...
+call(::BilinearDotFunctor, a, b) = a*b
+@inline bilindot{T <: Union{FixedArray, Tuple}}(a::T, b::T) = sum(map(DotFunctor(), a, b))
```
Comment (mschauer) [inline src/ops.jl]: Note #107

File: src/ops.jl [hunk 3]
```diff
@@ -115,21 +115,39 @@ end
...
+@inline bilindot{T1, T2}(a::NTuple{3,T1}, b::NTuple{3,T2}) = @inbounds return (a[1]*b[1] + a[2]*b[2] + a[3]*b[3])
+@inline bilindot{T1, T2}(a::NTuple{4,T1}, b::NTuple{4,T2}) = @inbounds return (a[1]*b[1] + a[2]*b[2] + a[3]*b[3]+a[4]*b[4])
+@inline bilindot{T1, T2, N}(a::NTuple{N,T1}, b::NTuple{N,T2}) = sum(map(DotFunctor(), promote(a, b)...))
+
+
 
 
 #cross{T}(a::FixedVector{2, T}, b::FixedVector{2, T}) = a[1]*b[2]-a[2]*b[1] # not really used!?
-@inline cross{T<:Number}(a::FixedVector{3, T}, b::FixedVector{3, T}) = @inbounds return typeof(a)(
+@inline cross{T1 <: Number, T2 <: Number}(a::FixedVector{3, T1}, b::FixedVector{3, T2}) = @inbounds return Vec(
```
Comment (c42f) [inline src/ops.jl]: This shouldn't really return a `Vec`, as `a` and `b` could be any user-defined type.

What type should it return?  I guess `similar(typeof(a), promote_type(T1,T2))` would be somewhat appropriate, though not really guaranteed to make sense for all user defined FSAs.  (And what about `b` anyway?  Well `Base` prefers the type of the left hand argument when deducing the output container to pass to `similar` for things like `+`, so I guess that'll do for now.)

**Comment by c42f** [comment]
@mschauer I don't think so.  I'm currently mostly messing with elementwise operations for which `map` is the right thing, but the matrix multiply stuff here seems largely independent.

File: src/ops.jl (hunk 3 above)
Comment (mschauer) [inline src/ops.jl]: Maybe a bit too much, but as the natural unit of the cross product vector is <s>volume</s> area and not length, something like http://docs.julialang.org/en/release-0.4/devdocs/promote-op/ or newer http://docs.julialang.org/en/latest/devdocs/promote-op/ might be possible. The reductions in base use this advance promotion, right?

Comment (c42f) [inline src/ops.jl]: True, `cross` should return something with volume units in principle.

`promote_op` is one solution, and I'm probably going to use that elsewhere but it's a bit dissatisfyingly manual.  The other would be a hypothetical `similar`-like function which constructs FSAs out of provided tuples.  This way, the element type can be inferred from the result type of expressions such as `a[2]*b[3]-a[3]*b[2]` - like is being done here with the `Vec` constructor - but without hardcoding the output type.

``` julia
# Possible (untested!) default, not sure about the name...
construct_similar{F<:FixedArray}(a::F, elements::Tuple) = basetype(F)(promote_type(elements...))
```

Comment (andyferris) [inline src/ops.jl]: Do we mean _area_ units? The cross product gives the area of a parallelogram. Take a further dot product with a length vector to get a volume.

Comment (mschauer) [inline src/ops.jl]: Yes, directed area, not directed volume! Thank you.

Comment (c42f) [inline src/ops.jl]: Eh, I copied that one right over without even noticing, quite right it's the area.

**Comment by awbsmith** [comment]
Hi all, I've:

1) made the line space change
2) changed the cross product to return `similar(typeof(a), promote_type(T1,T2))` as per @c42f https://github.com/c42f's suggestion. I don't think it's practical to know the return type, e.g. if its `cross(displacement, force)` you'd need a type defined for torque.  And returning the type of `a` will probably be less annoying than returning a generic Vec.  The good news is the change to the upstream master have made similar type safe.
3) rebased for the changes to the upstream master (and made a mess of it requiring a git history "fix", so sorry about the comment links)

Cheers,
Andrew

**Comment by c42f** [comment]
Thanks Andrew.  This is now superseded by #119 which also includes the changes here.

**Pull request closed without merging**